### PR TITLE
Updates

### PR DESCRIPTION
**Pull request by phadej**
A seed, so this repository doesn't look empty.

I'd also ditch `servant-contrib-client` and `servant-contrib-server` as I don't really see a point. New combinator forces to depend on all of them.

### CODE REVIEW

**Comment by alpmestan** [comment]
Well, if you only need the server side functionality, you probably want to depend only on the -server contrib package don't you?

**Comment by alpmestan** [comment]
LGTM

**Comment by fizruk** [comment]
I'd argue that this is not a servant combinator, just a new data type. Still useful and I'm happy to have it here.

`servant-contrib-client` and `servant-contrib-server` make sense when you actually introduce a new servant combinator with `HasClient` and `HasServer` instances. Dragging client dependencies into a server package is more or less ok, but dragging server deps into a client package (such as `telegram-api` or `servant-github`) is not good. Hence the necessary split of contrib packages.

File: servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+--
+-- @
+-- type API = "data" :> Capture "city" (RE.RE "a-zA-Z+") :> Get '[JSON] SomeData
+-- @
+module Servant.Contrib.RegexHttpApiData (
+    RE,
+    mk,
+    original,
+    groups,
+    ) where
```
Comment (fizruk) [inline servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs:25]: If you're not going to export everything, please use an `.Internal` module which does so.

File: servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+-- >>> original (unsafeMk "foobar" :: RE "o")
+-- "foobar"
+--
+-- prop> t == original (unsafeMk t :: RE ".*")
+original :: RE re -> T.Text
+original (RE t _) = t
+
+-- | Extract matching groups
+--
+-- >>> groups (unsafeMk "foobooo" :: RE "o+")
```
Comment (fizruk) [inline servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs:86]: You have examples and properties with `unsafeMk` which is not exported. More reason to use an `.Internal` module.

Or maybe there's a better way to write examples? How about something like this?

```
>>> groups <$> (mk "foobooo" :: RE "o+")
Just [["oo"],["ooo"]]
```

File: servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+-- | Partial 'mk'
+unsafeMk :: KnownSymbol re => T.Text -> RE re
+unsafeMk = fromJust . mk
+
+-- | Extract original string.
+--
+-- >>> original (unsafeMk "foobar" :: RE "o")
+-- "foobar"
+--
+-- prop> t == original (unsafeMk t :: RE ".*")
```
Comment (fizruk) [inline servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs:80]: Maybe rewrite this property without `unsafeMk`?

```
prop> pure t == original <$> (mk t :: RE ".*")
```

File: servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+-- Nothing
+--
+-- >>> mk "abc" :: Maybe (RE "foo")
+-- Nothing
+data RE (re :: Symbol) = RE !T.Text [[T.Text]]
+  deriving (Show, Typeable)
+
+-- | Construct 'RE' value.
+-- If regexp is invalid, or value doesn't match it 'Nothing' is returned.
+mk :: forall re. KnownSymbol re => T.Text -> Maybe (RE re)
```
Comment (fizruk) [inline servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs:61]: Can we distinguish "invalid regexp" and "value doesn't match it" cases?
Can we detect "invalid regexp" at compile time?

File: servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+-- [["oo"],["ooo"]]
+groups :: RE re -> [[T.Text]]
+groups (RE _ gs) = gs
+
+-------------------------------------------------------------------------------
+-- http-api-data
+-------------------------------------------------------------------------------
+
+instance KnownSymbol re => FromHttpApiData (RE re) where
+    parseUrlPiece = maybe (Left "doesn't match regex") Right . mk
```
Comment (fizruk) [inline servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs]: Can we also add the regex in question to the error message?

**Review by fizruk — CHANGES_REQUESTED** [review]
Apart from a few comments this looks good to me!

File: servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs (hunk 4 above)
Comment (phadej) [reply servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs:61]: Can we detect "invalid regexp", unfortunately no (without TH).

File: servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs (hunk 2 above)
Comment (phadej) [reply servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs:86]: Ah, forgot to export `unsafeMk`.

File: servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs (hunk 5 above)
Comment (phadej) [reply servant-contrib/src/Servant/Contrib/RegexHttpApiData.hs]: We can! good idea.

**Comment by phadej** [comment]
@fizruk yes, it could live and be useful outside `servant`, but I see `servant-contrib` where, though having quality requirements, aren't too strict about what belongs here and what not. More like collection of examples. After all, I had made opinionated choice to use `regex-tdfa`.

**Comment by phadej** [comment]
About combinators, so is the idea that combinator is defined in `servant-contrib` but `HasClient` and `HasServer` instances for it in `servant-contrib-client` and `servant-contrib-server` respectively. I don't like orphans, but indeed, I don't see better solution if one don't want to pull server deps to the client and vice versa.

**Review by fizruk — APPROVED** [review]

**Comment by fizruk** [comment]
Ok, I'm fine with the current state of this PR. Nice to have `servant-contrib` started!

> About combinators, so is the idea that combinator is defined in servant-contrib but HasClient and HasServer instances for it in servant-contrib-client and servant-contrib-server respectively. I don't like orphans, but indeed, I don't see better solution if one don't want to pull server deps to the client and vice versa.

Exactly, we haven't come up with any viable alternative when discussed options ~year ago. The only alternative I recall was to use cabal flags.

**Comment by alpmestan** [comment]
Yeah, until someone comes up with a better strategy, I think we can go with this one and see how it goes.

**Comment by phadej** [comment]
OK, I'll write those guidelines into README.md, PR will follow. 

(I cannot merge this)

**Pull request merged**